plugins/autoclothing.cpp

```diff
@@ -0,0 +1,643 @@
+#include "Core.h"
+#include <Console.h>
+#include <Export.h>
+#include <PluginManager.h>
+
+#include <map>
+
+// DF data structure definition headers
+#include "DataDefs.h"
+
+#include "modules/Items.h"
+#include "modules/Maps.h"
+#include "modules/Materials.h"
+#include "modules/Units.h"
+#include "modules/World.h"
+
+#include "df/itemdef_armorst.h"
+#include "df/itemdef_glovesst.h"
+#include "df/itemdef_shoesst.h"
+#include "df/itemdef_helmst.h"
+#include "df/itemdef_pantsst.h"
+#include "df/manager_order.h"
+#include "df/creature_raw.h"
+#include "df/world.h"
+
+using namespace DFHack;
+using namespace DFHack::Items;
+using namespace DFHack::Units;
+using namespace df::enums;
+
+
+// A plugin must be able to return its name and version.
+// The name string provided must correspond to the filename -
+// skeleton.plug.so, skeleton.plug.dylib, or skeleton.plug.dll in this case
+DFHACK_PLUGIN("autoclothing");
+
+// Any globals a plugin requires (e.g. world) should be listed here.
+// For example, this line expands to "using df::global::world" and prevents the
+// plugin from being loaded if df::global::world is null (i.e. missing from symbols.xml):
+//
+REQUIRE_GLOBAL(world);
+
+// Only run if this is enabled
+DFHACK_PLUGIN_IS_ENABLED(autoclothing_enabled);
+
+// Here go all the command declarations...
+// mostly to allow having the mandatory stuff on top of the file and commands on the bottom
+struct ClothingRequirement;
+command_result autoclothing(color_ostream &out, std::vector <std::string> & parameters);
+static void init_state(color_ostream &out);
+static void save_state(color_ostream &out);
+static void cleanup_state(color_ostream &out);
+static void do_autoclothing();
+static bool validateMaterialCategory(ClothingRequirement * requirement);
+static bool setItem(std::string name, ClothingRequirement* requirement);
+
+std::vector<ClothingRequirement>clothingOrders;
+
+struct ClothingRequirement
+{
+    df::job_type job_type;
+    df::item_type item_type;
+    int16_t item_subtype;
+    df::job_material_category material_category;
+    int16_t needed_per_citizen;
+    std::map<int16_t, int32_t> total_needed_per_race;
+
+    bool matches(ClothingRequirement * b)
+    {
+        if (b->job_type != this->job_type)
+            return false;
+        if (b->item_type != this->item_type)
+            return false;
+        if (b->item_subtype != this->item_subtype)
+            return false;
+        if (b->material_category.whole != this->material_category.whole)
+            return false;
+        return true;
+    }
+
+    std::string Serialize()
+    {
+        stringstream stream;
+        stream << job_type << " ";
+        stream << item_type << " ";
+        stream << item_subtype << " ";
+        stream << material_category.whole << " ";
+        stream << needed_per_citizen;
+        return stream.str();
+    }
+
+    void Deserialize(std::string s)
+    {
+        stringstream stream(s);
+        stream >> (int16_t&)job_type;
```

Comment:
This is throwing a warning (see buildmaster). I personally don't think it should throw a warning, but it is. Maybe a better solution is defining a custom operator>> for job_type and item_type, although that should probably be done for all DF enums if we're going that route.

Reply:
I changed it to serialize the enums to string instead of to numbers. It won't be as fast, but it's only when saving and loading, so it should be fine.

Reply:
Cool, thanks for that. Wish the original way had worked... might look into adding I/O operators anyway.